### DISCUSSION TITLE
A (hopefully) simple SyGuS problem: setting a bit inside a byte

### DISCUSSION THREAD

**Discussion by jefftrull** [cvc5/cvc5#10435] [Q&A] [answered]
I've successfully synthesized a few programs using SyGuS but am now stuck. My code seems right to me but it doesn't result in a successful synthesis. I have reduced it to a simple example of setting a bit inside a byte. My constraint is set up to verify that bit 7 is set:

```scheme
;; just verify correctness for bit 7
(constraint
 (let ((result (tbd x y)))
   (and
    (= ((_ extract 7 7) result) y)
    (= ((_ extract 6 0) result) ((_ extract 6 0) x))
)))
```

I have a separate `bit-setter` function that accepts a bit number (0-7), the original byte, and the new value for the chosen bit. I use it in a grammar like this:

```scheme
(synth-fun tbd ((x Byte) (y Bit)) Byte
           ((Start Byte) (bitno (_ BitVec 3)))
           ((Start Byte (
                         x
                         (bit-setter bitno y x)
                         ))
            (bitno (_ BitVec 3) ((Constant (_ BitVec 3))))
))
```

I expect it to synthesize `(bit-setter #b111 y x)` but it succeeds only for the most trivial implementations of `bit-setter` and does not (indeed, cannot) choose `bitno`. If I allow a recursive grammar here it synthesizes `(bit-setter #b111 y (bit-setter #b111 y x))` which _really_ doesn't make sense (IMO).  Full code:

```scheme
(set-logic BV)
(set-option :produce-models true)

(define-sort Byte () (_ BitVec 8))
(define-sort Bit () (_ BitVec 1))

;; this "works" for the case of bit 7 but does not chose bitno
;; (define-fun bit-setter ((bitno (_ BitVec 3)) (y Bit) (x Byte)) Byte
;;             ;; only correct only for bitno=#b111
;;             (concat y ((_ extract 6 0) x))

;; this does not succeed, even though we are synthesizing for bit 7
;; surely it would be easy to select bitno=#b111 here
(define-fun bit-setter ((bitno (_ BitVec 3)) (y Bit) (x Byte)) Byte
            ;; correct for bitno=#b111 and bitno=#b110
            (ite (= bitno #b111)
                 (concat y ((_ extract 6 0) x))
                 (concat ((_ extract 7 7) x) y ((_ extract 5 0) x))))

;; a reduced grammar demonstrating the problem
(synth-fun tbd ((x Byte) (y Bit)) Byte
           ((Start Byte) (bitno (_ BitVec 3)))
           ((Start Byte (
                         x
                         (bit-setter bitno y x)
                         ))
            (bitno (_ BitVec 3) ((Constant (_ BitVec 3))))
))

(declare-var x Byte)
(declare-var y Bit)

;; just verify correctness for bit 7
(constraint
 (let ((result (tbd x y)))
   (and
    (= ((_ extract 7 7) result) y)
    (= ((_ extract 6 0) result) ((_ extract 6 0) x))
)))

(check-synth)
```

**Comment by ajreynol** [comment, marked as answer]
Thanks for the example.  cvc5 fails to solve this due to an incomplete heuristic for how we handle constant production rules `(Constant (_ BitVec 3))`.  

Using `--sygus-repair-const` allows cvc5 to quickly solve with this solution:
```
(
(define-fun tbd ((x (_ BitVec 8)) (y (_ BitVec 1))) (_ BitVec 8) (bit-setter #b111 y x))
)
```

We've seen incompleteness issues like this before. I am going to change our default options to enable `--sygus-repair-const` by default.

**Reply by jefftrull** [reply]
Perfect, that fixes the script I extracted the test case from as well. Thanks! PS: is there a performance cost to this?

**Comment by ajreynol** [comment]
In general, yes, this technique invokes a "heavier hammer" internally. It relies on a call to a subsolver for exists-forall solving for certain enumerated candidate solutions.  Details can be found in this paper: https://homepage.divms.uiowa.edu/~ajreynol/jar2023a.pdf

Nevertheless, this technique will clearly scale better than a brute force enumeration of possible constants, e.g. when the bitwidth of the constant you are looking for is large.

Do feel free to send more challenge benchmarks if you run into issues.
